nengo_spa/compiler/tests/test_ast.py

```diff
@@ -0,0 +1,436 @@
+from io import BytesIO
+
+import pytest
+
+import nengo_spa as spa
+from nengo_spa.exceptions import SpaTypeError
+from nengo_spa.compiler import tokens
+from nengo_spa.compiler.ast import AstBuilder
+from nengo_spa.compiler.ast_nodes import (
+    Action, ActionSet, ApproxInverse, DotProduct, Effect, Effects, Negative,
+    Module, Product, Scalar, Sink, Sum, Symbol)
+from nengo_spa.compiler.ast_types import (
+    TActionSet, TEffect, TEffects, TScalar, TVocabulary)
+
+
+def bare_tokens(expr):
+    return [
+        tk for tk in tokens.tokenize(BytesIO(expr.encode()).readline)
+        if tk[0] != tokens.ENCODING and tk[0] != tokens.ENDMARKER]
+
+
+def test_scalar():
+    ast = AstBuilder().build_expr(bare_tokens('1'))
+    assert ast == [1]
+
+
+def test_symbol():
+    ast = AstBuilder().build_expr(bare_tokens('A'))
+    assert ast == [Symbol('A')]
+    node = ast[0]
+    assert str(node) == 'A'
+
+    vocab_type = TVocabulary(spa.Vocabulary(16, strict=False))
+    node.infer_types(None, vocab_type)
+    assert node.type == vocab_type
+
+    ast = AstBuilder().build_expr(bare_tokens('A'))
+    with pytest.raises(SpaTypeError):
+        ast[0].infer_types(None, TScalar)
+
+
+def test_spa_network():
+    d = 16
+    with spa.Network() as model:
+        state = spa.State(d)
+
+    ast = AstBuilder().build_expr(bare_tokens('state'))
+    assert ast == [Module('state', state)]
+    assert str(ast[0]) == 'state'
+    ast[0].infer_types(model, None)
+    assert ast[0].type.vocab == state.vocabs[d]
+
+    with spa.Network() as model:
+        with spa.Network() as network:
+            network.state = spa.State(d)
+
+    ast = AstBuilder().build_expr(bare_tokens('network.state'))
+    assert ast == [Module('network.state', network.state)]
+    assert str(ast[0]) == 'network.state'
+    ast[0].infer_types(model, None)
+    assert ast[0].type.vocab == network.state.vocabs[d]
+
+
+def test_scalar_multiplication():
```

Comment:
I feel like this could be broken up into a separate tests. One that tests inferring vocabulary directly and one that tests inferring vocabulary from the network.

Reply:
These tests have just been moved to a new place. Which doesn't mean that the structure couldn't be improved, but if we split up this test, it should be done to all other tests too. Right now there is one test function corresponding to each type of AST node and they all have a similar structure.